pkg/tools/clientmap/refcount.go

```diff
@@ -52,12 +49,21 @@ func (r *RefcountMap) Store(key string, value networkservice.NetworkServiceClien
 // loaded result is true if the value was loaded, false if stored.
 // Increments the refcount.
 func (r *RefcountMap) LoadOrStore(key string, value networkservice.NetworkServiceClient) (networkservice.NetworkServiceClient, bool) {
-	r.lock.RLock()
-	defer r.lock.RUnlock()
-
-	raw, loaded := r.m.LoadOrStore(key, &refcountClient{NetworkServiceClient: value})
+	raw, loaded := r.m.LoadOrStore(key, &refcountClient{
+		count:                1,
+		NetworkServiceClient: value,
+	})
 	client := raw.(*refcountClient)
+	if !loaded {
+		return client.NetworkServiceClient, false
+	}
 
+	client.lock.RLock()
+	defer client.lock.RUnlock()
+
+	if atomic.LoadInt32(&client.count) <= 0 {
```

Comment:
Do we still need to use the atomic under the locker?

Reply:
Yep, it is `RLock()` not `Lock()`.

Reply:
RLock means that you could safely _read_. So I see that it is still over synchronized.

Reply:
`RLock`, `Lock` here are not for reading and writing `client.count`. `RLock` is for updating `client.count`. `Lock` is for deleting `client`.
`RLock` doesn't guarantee that only 1 goroutine updates `client.count` at the moment and it is actually what we need for the updates to increase performance.
https://github.com/networkservicemesh/sdk/blob/fe5ced750db9551ac39a3780e494e4b3fed76154/pkg/tools/clientmap/refcount.go#L61-L67
`Lock` guarantees that no one would restore `client` after it as been deleted and so it does some guarantees:
https://github.com/networkservicemesh/sdk/blob/fe5ced750db9551ac39a3780e494e4b3fed76154/pkg/tools/clientmap/refcount.go#L113-L125

Reply:
Reworked with simple `sync.Mutex` in another PR: https://github.com/networkservicemesh/sdk/pull/610.